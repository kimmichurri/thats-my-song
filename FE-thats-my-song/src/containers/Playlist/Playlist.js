import React, { Component } from 'react';
import { NavLink } from 'react-router-dom';
import { connect } from 'react-redux';
import PropTypes from 'prop-types';
import { getFromPlaylist } from '../../thunks/getFromPlaylist';
import PlaylistCard from '../../components/PlaylistCard/PlaylistCard';
import { deleteSong } from '../../thunks/deleteSong';
import { deleteFromPlaylist } from '../../actions';

export class Playlist extends Component {

  goToPlaylist = () => {
    const url = 'http://localhost:3001/api/v1/playlist/';
    this.props.getFromPlaylist(url);
  }

  componentDidMount = () => {
   this.goToPlaylist();
  }

  deleteSongFromPlaylist = (e) => {
    const id = e.target.value
    const numericId = parseInt(e.target.value);
    const url = `http://localhost:3001/api/v1/playlist/${id}`;
    this.props.deleteSong(url, id);
    this.props.deleteFromPlaylist(numericId);
  }

  render() {
    let playlistCards = this.props.playlist.map(song => {
      return (
        <PlaylistCard
          key={song.id}
          id={song.id}
          title={song.title}
          artist={song.artist}
          deleteSongFromPlaylist={this.deleteSongFromPlaylist}
        />
      )
    });

    return (
      <div className='playlist-wrapper'>
        <h2 className='playlist-title'>Welcome to your Playlist!</h2>
        <div className='playlist-songs-cards-display'>
          {playlistCards}
        </div>
        <NavLink to='/' id='return-to-search-navlink'>
          <button className='song-info-page-buttons'>Return to Search</button>
      </NavLink> 
      </div>
    )
  }
}

Playlist.propTypes = {
  playlist: PropTypes.array.isRequired,
  getFromPlaylist: PropTypes.func.isRequired
}

export const mapStateToProps = (state) => ({
  playlist: state.setPlaylist,
});

export const mapDispatchToProps = (dispatch) => ({
  getFromPlaylist: (url) => dispatch(getFromPlaylist(url)),
  deleteSong: (url, id) => dispatch(deleteSong(url, id)),
  deleteFromPlaylist: (id) => dispatch(deleteFromPlaylist(id))
});

export default connect(mapStateToProps, mapDispatchToProps)(Playlist);