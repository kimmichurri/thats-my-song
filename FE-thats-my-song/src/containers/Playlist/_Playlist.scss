.playlist-title {
  font-family: $main-font;
  font-size: 2.5rem;
}

.playlist-wrapper {
  display: flex;
  flex-direction: column;
  text-align: center;
}

.playlist-songs-cards-display {
  display: flex;
  flex-wrap: wrap;
  padding: 0% 10% 5%;
}