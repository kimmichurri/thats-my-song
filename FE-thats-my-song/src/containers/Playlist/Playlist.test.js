import  { Playlist, mapStateToProps, mapDispatchToProps } from './Playlist';
import React from 'react';
import { shallow } from 'enzyme';
jest.mock('../../thunks/getFromPlaylist');

describe('Playlist', () => {
  let wrapper;
  const mockProps = {
    playlist: [{title: 'a song', artist: 'kim', id: 1}],
    getFromPlaylist: jest.fn()
  } 

  beforeEach(() => {
    wrapper = shallow(
      <Playlist 
        playlist={mockProps.playlist}
        getFromPlaylist={mockProps.getFromPlaylist}
      />
    )
  });

  it('should match the snapshot', () => {
    expect(wrapper.debug()).toMatchSnapshot();
  });

  describe('mapStateToProps', () => {
    it('should return a playlist of songs', () => {
      const mockState = {
        randomThing: 1,
        anotherThing: 'two',
        setPlaylist: [{song: 'a'}, {song: 'b'}]
      }
      const expected = {
        playlist: [{song: 'a'}, {song: 'b'}]
      }
      const mappedProps = mapStateToProps(mockState);
      expect(mappedProps).toEqual(expected);
    });
  });

  describe('mapDispatchToProps', () => {
    it('should return getFromPlaylist to dispatch', () => {
      const mockDispatch = jest.fn();
      const getFromPlaylist = jest.fn();
      const actionToDispatch = getFromPlaylist('www.playlist.com');
      const mappedProps = mapDispatchToProps(mockDispatch);
      mappedProps.getFromPlaylist('www.playlist.com')
      expect(mockDispatch).toHaveBeenCalledWith(actionToDispatch);
    });
  });
});