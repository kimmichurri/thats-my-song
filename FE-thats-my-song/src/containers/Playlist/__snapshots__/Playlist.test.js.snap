// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Playlist should match the snapshot 1`] = `
"<div className=\\"playlist-wrapper\\">
  <h2>
    Welcome to your Playlist!
  </h2>
  <PlaylistCard id={1} title=\\"a song\\" artist=\\"kim\\" />
  <NavLink to=\\"/\\" id=\\"return-to-search-navlink\\">
    <button className=\\"song-info-page-buttons\\">
      Return to Search
    </button>
  </NavLink>
</div>"
`;
